File: Google-Kick-Start/2019/Kick_Start_2019-E-1.cpp
// Cherries Mesh
// https://codingcompetitions.withgoogle.com/kickstart/round/0000000000050edb/0000000000170721

#include <bits/stdc++.h>

typedef long long ll;

using namespace std;

int main() {
    int T;
    cin >> T;
    for (int t = 0; t < T; t++) {
        int N, M;
        cin >> N >> M;
        vector< set<int> > graph(N);
        for (int i = 0; i < M; i++) {
            int c, d;
            cin >> c >> d;
            graph[c - 1].insert(d - 1);
            graph[d - 1].insert(c - 1);
        }
        
        set<int> unchecked;
        for (int i = 0; i < N; i++) {
            unchecked.insert(i);
        }
        int group_count = 0;
        while (!unchecked.empty()) {
            set<int> queue;
            queue.insert(*unchecked.begin());
            while (!queue.empty()) {
                int v = *queue.begin();
                queue.erase(v);
                unchecked.erase(v);
                for (auto w : graph[v]) {
                    if (unchecked.find(w) != unchecked.end()) {
                        queue.insert(w);
                    }
                }
            }
            group_count += 1;
        }

        int result = (group_count - 1) * 2 + (N - group_count);
        printf("Case #%d: %d\n", t + 1, result);
    }
}

File: Google-Kick-Start/2019/Kick_Start_2019-E-1.py
# Cherries Mesh
# https://codingcompetitions.withgoogle.com/kickstart/round/0000000000050edb/0000000000170721

T = int(input())
for t in range(T):
    [N, M] = [int(n) for n in input().split()]
    graph = [set() for _ in range(N)]
    for i in range(M):
        [c, d] = [int(n) for n in input().split()]
        graph[c - 1].add(d - 1)
        graph[d - 1].add(c - 1)

    unchecked = set(range(N))
    group_count = 0
    while unchecked:
        queue = set([next(iter(unchecked))])
        while queue:
            # print(queue)
            v = next(iter(queue))
            queue.remove(v)
            unchecked.remove(v)
            for w in graph[v]:
                if w in unchecked and w not in queue:
                    queue.add(w)
        group_count += 1

    ans = (group_count - 1) * 2 + (N - group_count)
    print(f"Case #{t + 1}: {ans}")
